### PR TITLE
Add wolfCrypt port for the Nuvoton NuMicro M2354

### PR DESCRIPTION
**Pull request by dgarske**
## Summary

The NuMicro M2354 is an Arm Cortex-M23 (Armv8-M baseline) part with TrustZone, a CRPT cryptographic accelerator, a separate TRNG and a Key Store holding keys in SRAM, Flash and OTP slots. This adds a crypto callback port for it, so software stays compiled in and anything the accelerator cannot do is declined rather than failed.

**What it adds**
- Hashing on the engine: SHA-1, SHA-224, SHA-256, SHA-384 and SHA-512, streaming through the engine's feedback state so a context costs a fixed amount rather than growing with the message.
- AES on the engine: ECB, CBC and CTR, plus **AES-GCM** and **AES-CCM** authenticated encryption. A GCM message of any size is carried through the engine's DMA cascade.
- Public key on the engine: ECDSA sign and verify, ECDH and key generation over P-192 through P-521 and Brainpool, and RSA from 1024 to 4096 bits.
- TRNG seeding, wired into the `wc_GenerateSeed()` chain so a `WC_RNG` built with `INVALID_DEVID` still gets entropy on bare metal.
- Key Store support: an AES key can live in an SRAM, Flash or OTP slot and be used by handle, so the key material never enters wolfSSL memory.
- Either TrustZone world. The same `wc_nuvoton_hw_*` interface calls the BSP directly from the secure world, or resolves through `cmse_nonsecure_entry` veneers from the non-secure world.
- A cross-compile CI job building both worlds against the real vendor BSP at a pinned commit.

**Benchmarks.** NuMaker-M2354 at 96 MHz. Both columns come from the same build, with SP ECC/RSA and the Thumb-1 assembly enabled, so the software column is wolfSSL configured properly for this core rather than falling back to portable C.

| Algorithm | Software | Hardware |
|---|---|---|
| SHA-1 | 1.850 MiB/s | 8.521 MiB/s |
| SHA-256 | 771.1 KiB/s | 8.707 MiB/s |
| SHA-512 | 270.9 KiB/s | 13.951 MiB/s |
| HMAC-SHA-512 | 266.5 KiB/s | 12.671 MiB/s |
| AES-128-CBC enc | 411.4 KiB/s | 8.732 MiB/s |
| AES-256-CBC enc | 293.5 KiB/s | 7.837 MiB/s |
| AES-128-CTR | 435.6 KiB/s | 8.561 MiB/s |
| AES-128-ECB enc | 451.0 KiB/s | 11.494 MiB/s |

The port README carries the full table.

## Hardware / test status

Validated on a NuMaker-M2354 (M2354KJFAE at 96 MHz, Nu-Link2-Me CMSIS-DAP, UART0 on PA6/PA7 at 115200), flashed with `pyocd` against its builtin `m2354kjfae` target. `wolfcrypt_test` completes with **42 tests passing and none failing**, covering every offloaded engine.

Two paths the suite does not reach were covered separately on the board: a Key Store round trip, and a sweep of AES-GCM and AES-CCM against software across payload, nonce and tag lengths.

The non-secure TrustZone world was also run on the board, reaching the accelerator through the veneers. The port README records what passed there, and one behaviour in a combined run that is not yet understood.

Host `--enable-all --enable-cryptocb` `make check` is unchanged, `check-headers` and `check-source-text` are clean, and both cross-compile legs build under the project warning set.

## Related

wolfBoot's M2354 target builds on this port: https://github.com/wolfSSL/wolfBoot/pull/884

### CODE REVIEW

File: wolfcrypt/src/port/nuvoton/nuvoton_hw.c [hunk 1]
```diff
@@ -0,0 +1,1924 @@
...
+
+        if (oneShot) {
+            ret = nuvoton_gcm_run(ctl, nuvotonGcmIn, nuvotonGcmOut, packedSz);
+        }
+        else {
+            ret = nuvoton_gcm_cascade(req, ctl);
+        }
+    }
+
+    if (ret == 0) {
```
Comment (aidangarske) [inline wolfcrypt/src/port/nuvoton/nuvoton_hw.c:1254]: **GCM tag repair breaks valid in-place decrypts**

The tag-repair path runs after the hardware result has been copied to `req->out`. For decrypts it then passes `req->in` to `nuvoton_gcm_tag()`. When `in == out`, which wolfCrypt explicitly supports (`tests/api/test_aes.c:3527-3585`), that pointer now contains plaintext rather than the ciphertext GCM must authenticate. Valid in-place decrypts with payload lengths congruent to 1 or 15 modulo 16 therefore fail authentication and `wc_NuvotonAeadDecFinish()` wipes the plaintext. For cascade-sized requests, a later `BAD_LENGTH_E` is translated to `CRYPTOCB_UNAVAILABLE` after output has already been changed, so the software fallback also operates on altered in-place input.

**Fix:** Preserve the original ciphertext for decrypt tag repair, or decline these requests before modifying caller buffers. Preflight cascade repair capacity before doing any output write. Add on-target tests for 1-byte, 15-byte, and cascade-sized in-place GCM encrypt/decrypt requests.

File: wolfcrypt/src/port/nuvoton/nuvoton_key.c [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+        return BAD_FUNC_ARG;
+    }
+    if (ksKey->owner != WC_NUVOTON_KS_OWNER_AES) {
+        return BAD_FUNC_ARG;
+    }
+    if (ksKey->bits != 128 && ksKey->bits != 192 && ksKey->bits != 256) {
+        return BAD_FUNC_ARG;
+    }
+
+    /* Any key this Aes already held has to go. Leaving the software schedule
```
Comment (aidangarske) [inline wolfcrypt/src/port/nuvoton/nuvoton_key.c:133]: **Raw rekey still selects the prior Key Store key**

Attaching a Key Store key leaves its handle in `aes->devCtx`. A later normal `wc_AesSetKey()` copies the new raw key into `aes->devKey` but does not clear `devCtx`. The new cipher callback gives `devCtx` precedence (`nuvoton_cb_cipher.c:104-117`), so the subsequent AES operation silently uses the old Key Store slot instead of the newly supplied raw key. Reusing an initialized `Aes` object in this order therefore produces output under the wrong key.

**Fix:** Define a raw-key transition that clears the Nuvoton Key Store association before software/raw-key setup, such as a Nuvoton set-key callback. Add regression coverage that attaches a Key Store key, calls `wc_AesSetKey()` with a different key, and verifies CBC, CTR, and GCM use the new key.

File: wolfcrypt/src/asn.c [hunk 3]
```diff
@@ -33997,17 +33997,22 @@ int DecodeECC_DSA_Sig_Ex(const byte* sig, word32 sigLen, mp_int* r, mp_int* s,
 #endif
 
 
-#ifdef WOLFSSL_ASN_TEMPLATE
-#if defined(HAVE_ECC) && defined(WOLFSSL_CUSTOM_CURVES)
+#ifdef WOLFSSL_ASN_HEX_STRING
```
Comment (aidangarske) [inline wolfcrypt/src/asn.c]: **NO_ASN public-key builds leave Nuvoton symbols unavailable**

`nuvoton_settings.h` enables `WOLFSSL_ASN_HEX_STRING` whenever the ECC or RSA port is selected, including `NO_ASN` configurations. The declaration is outside the `NO_ASN` guard in `asn.h`, but this new definition is inside `asn.c`'s enclosing `#ifndef NO_ASN` block (`asn.c:266-40759`). Thus an RSA-only `NO_ASN` build compiles a call from `nuvoton_cb_pk.c:800` without a definition at link time. With ECC enabled, the new sign and verify paths additionally reference `StoreECC_DSA_Sig()` and `DecodeECC_DSA_Sig()`, which are also excluded by `NO_ASN`.

**Fix:** Either support `NO_ASN` by moving the standalone hex helper outside the ASN guard and declining/adapting DER-dependent ECDSA paths, or reject and document the unsupported Nuvoton ECC/RSA plus `NO_ASN` configuration. Add a crypto-only cross-build to CI.

File: .github/workflows/nuvoton-m2354-compile.yml [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+# The runnable example, including those veneers, lives in wolfssl-examples
+# under embedded/nuvoton_m2354. Functional correctness is validated on a
+# NuMaker-M2354, not here; see wolfcrypt/src/port/nuvoton/README.md.
+
+# START OF COMMON SECTION
+on:
+  push:
+    branches: [ 'master', 'main', 'release/**' ]
+    paths:
+      - 'wolfcrypt/src/port/nuvoton/**'
```
Comment (aidangarske) [inline .github/workflows/nuvoton-m2354-compile.yml]: **Post-merge core changes skip the new compile guard**

The pull-request filter covers all wolfCrypt sources and headers, but the push filter covers only the Nuvoton directories. A direct change on `master`, `main`, or a release branch to a dependency such as `aes.c`, `ecc.c`, `random.c`, or `cryptocb.h` will not run this guard, despite the workflow stating that it prevents such refactors from silently breaking the port.

**Fix:** Use the same core source/header path coverage for push events as for pull requests, or remove the push path filter.

**Suggestion:**
```suggestion
- 'wolfcrypt/src/**'
      - 'wolfssl/wolfcrypt/**'
      - '.github/workflows/nuvoton-m2354-compile.yml'
```

**Review by aidangarske — CHANGES_REQUESTED** [review]
## Skoll Code Review

**Scan type:** `review`

**Overall recommendation:** `REQUEST_CHANGES`
**Findings:** 4 total — 4 posted, 0 skipped
**4 finding(s) posted as inline comments** (see file-level comments below)

### Posted findings
- [High] **GCM tag repair breaks valid in-place decrypts** — `wolfcrypt/src/port/nuvoton/nuvoton_hw.c:984-1001`
- [High] **Raw rekey still selects the prior Key Store key** — `wolfcrypt/src/port/nuvoton/nuvoton_key.c:125-142`
- [High] **NO_ASN public-key builds leave Nuvoton symbols unavailable** — `wolfcrypt/src/asn.c:34000-34029`
- [Medium] **Post-merge core changes skip the new compile guard** — `.github/workflows/nuvoton-m2354-compile.yml:27-29`

*Review generated by [Skoll](https://github.com/wolfSSL/fenrir/tree/main/skoll)*

File: wolfcrypt/src/port/nuvoton/nuvoton_key.c (hunk 2 above)
Comment (aidangarske) [inline wolfcrypt/src/port/nuvoton/nuvoton_key.c:133]: **Stored-key binding survives a subsequent raw rekey**

`wc_NuvotonKs_SetAesKey()` binds the Key Store handle in `aes->devCtx`, but a later normal `wc_AesSetKey()` does not clear that field. The standard CryptoCB set-key path updates `aes->devKey`; the new cipher callback still sees non-NULL `devCtx`, selects `keySlot`, and ignores the newly supplied raw key. Reusing an `Aes` object as `SetAesKey(slotA) -> wc_AesSetKey(rawB) -> encrypt` therefore succeeds while encrypting with slotA, not rawB.

**Fix:** Define a safe unbind/rekey lifecycle: a raw `wc_AesSetKey()` must clear the Nuvoton stored-key binding before software or raw-key callback setup, or expose an explicit unbind operation and reject raw rekeying while bound. Also wipe stale `aes->devKey` when attaching a stored key.

File: wolfcrypt/src/port/nuvoton/nuvoton_hw.c [hunk 5]
```diff
@@ -0,0 +1,1924 @@
...
+{
+    uint32_t ctl;
+    uint32_t j0[4];
+    word32   aadSection;
+    word32   cSection;
+    word32   len;
+    word32   i;
+    int      ret;
+
+    aadSection = nuvoton_gcm_align(req->aadSz);
```
Comment (aidangarske) [inline wolfcrypt/src/port/nuvoton/nuvoton_hw.c:873]: **Large GCM operations fall back only after mutating an in-place buffer**

For payload lengths congruent to 1 or 15 modulo 16, the implementation recomputes the tag after the GCM cascade has written ciphertext/plaintext to `req->out`. That helper rejects a request once its GHASH input exceeds the fixed 1088-byte staging buffer. For example, a 1089-byte message with no AAD takes the cascade, writes the output, then returns `BAD_LENGTH_E` because the tag input is 1120 bytes. `wc_NuvotonAesRun()` converts that to `CRYPTOCB_UNAVAILABLE` for ordinary keys, causing the core to retry software GCM against an already overwritten `in == out` buffer. Encryption double-transforms the data; decryption authenticates the wrong input. Stored keys instead fail with `WC_HW_E` despite the advertised arbitrary-size GCM support.

**Fix:** Preflight this limitation before any output is written, or implement tag recomputation as a streaming/cascade GHASH path. Add boundary tests above the one-shot buffer size for payload remainders 1 and 15, including `in == out` and Key Store keys.

File: wolfcrypt/src/port/nuvoton/nuvoton_cb_pk.c [hunk 6]
```diff
@@ -0,0 +1,869 @@
...
+     * stands in for the whole of wc_ecc_sign_hash(). */
+    if (ret == 0) {
+        ret = mp_init_multi(r, s, NULL, NULL, NULL, NULL);
+        if (ret == MP_OKAY) {
+            ret = mp_read_radix(r, hex->r, MP_RADIX_HEX);
+            if (ret == MP_OKAY) {
+                ret = mp_read_radix(s, hex->s, MP_RADIX_HEX);
+            }
+            if (ret == MP_OKAY) {
+                ret = StoreECC_DSA_Sig(info->pk.eccsign.out,
```
Comment (aidangarske) [inline wolfcrypt/src/port/nuvoton/nuvoton_cb_pk.c:462]: **Public-key port does not build in valid no-ASN ECC/RSA configurations**

The port enables ECC/RSA callbacks regardless of `NO_ASN`, but its new ECDSA code always calls `StoreECC_DSA_Sig()` and `DecodeECC_DSA_Sig()`. Those helpers, along with the newly exported `wc_DataToHexString()` implementation used by RSA, are unavailable when `NO_ASN` and `NO_PWDBASED` are set. This is a supported wolfCrypt configuration: the software ECDSA path explicitly emits and accepts raw `r || s` signatures in that mode. Compiling the new callback with `NO_ASN`, `NO_PWDBASED`, ECC, and no DSA produces undeclared-function errors for both ASN signature helpers.

**Fix:** Make the public-key port ASN-independent, including raw `r || s` handling for no-ASN ECDSA and a non-ASN hex converter, or reject Nuvoton ECC/RSA offload under `NO_ASN` in `nuvoton_settings.h` with a clear configuration error.

File: .github/workflows/nuvoton-m2354-compile.yml [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+          EOF
+
+          CFLAGS="-mcpu=cortex-m23 -mthumb -Os -Wall -Wextra -Werror -c"
+          CFLAGS="$CFLAGS -I. -I$GITHUB_WORKSPACE/cfg -DWOLFSSL_USER_SETTINGS"
+          CFLAGS="$CFLAGS -D${{ matrix.world }}"
+          CFLAGS="$CFLAGS -I$BSP/Library/StdDriver/inc"
+          CFLAGS="$CFLAGS -I$BSP/Library/Device/Nuvoton/M2354/Include"
+          CFLAGS="$CFLAGS -I$BSP/Library/CMSIS/Include"
+
+          for f in wolfcrypt/src/port/nuvoton/*.c wolfcrypt/src/random.c; do
```
Comment (aidangarske) [inline .github/workflows/nuvoton-m2354-compile.yml:134]: **CI validates compilation only, not callback behavior**

The only in-tree Nuvoton coverage compiles each source file with `-c`; there are no Nuvoton test callers in the repository. It cannot detect callback dispatch, state handoff, Key Store lifecycle, AEAD fallback, or hash-copy regressions. The two blocking cases above would both pass this job.

**Fix:** Add runnable regression coverage, using board tests or a mock hardware layer, for callback fallback before output mutation, in-place GCM cascade boundaries, stored-key/raw-key transitions, and hash update/copy/free state.

**Review by aidangarske — CHANGES_REQUESTED** [review]
## Skoll Code Review

**Scan type:** `review`

**Overall recommendation:** `REQUEST_CHANGES`
**Findings:** 4 total — 4 posted, 0 skipped
**4 finding(s) posted as inline comments** (see file-level comments below)

### Posted findings
- [High] **Stored-key binding survives a subsequent raw rekey** — `wolfcrypt/src/port/nuvoton/nuvoton_key.c:125-142`
- [High] **Large GCM operations fall back only after mutating an in-place buffer** — `wolfcrypt/src/port/nuvoton/nuvoton_hw.c:730-735`
- [High] **Public-key port does not build in valid no-ASN ECC/RSA configurations** — `wolfcrypt/src/port/nuvoton/nuvoton_cb_pk.c:400-463`
- [Medium] **CI validates compilation only, not callback behavior** — `.github/workflows/nuvoton-m2354-compile.yml:131-134`

*Review generated by [Skoll](https://github.com/wolfSSL/fenrir/tree/main/skoll)*

**Review by aidangarske — APPROVED** [review]
